File: objc/Tests/TestTransportWriteDatasetPrelude.m
/*
 * TestTransportWriteDatasetPrelude.m — Task 3.9 of transport-spec v0.11.
 *
 * End-to-end verification of -[TTIOTransportWriter writeDataset:]'s
 * v0.11 prelude wiring. The prelude emits in this strict order per
 * spec §5.4:
 *
 *   §5.4.1 ENCRYPTION_ALGORITHM (0x1B)  — when ds.isEncrypted
 *   §5.4.2 DATASET_PROVENANCE   (0x18)  — when ds.provenanceRecords non-empty
 *   §5.4.3 SUBJECT_METADATA / SAMPLE_METADATA  — deferred
 *   §5.4.4 reference groups (0x10/0x11/0x12)   — one per ds.references value
 *   §5.4.5 image cube (0x13/0x14/0x15)         — when ds.msImage != nil
 *   §5.4.6 IDENTIFICATIONS_TABLE (0x16) / QUANTIFICATIONS_TABLE (0x17)
 *
 * The transport_v0_11 feature flag rides on StreamHeader.features when
 * any of the above is non-empty / non-nil.
 *
 * Cross-language parity:
 *   Java TransportWriterReferenceWireUpTest (commit dc0de926)
 *   Python tests/test_transport_write_dataset_prelude.py (commit 6f51e81b)
 *
 * SPDX-License-Identifier: Apache-2.0
 */
#import <Foundation/Foundation.h>
#import <objc/runtime.h>
#import "Testing.h"

#import "Transport/TTIOTransportPacket.h"
#import "Transport/TTIOTransportReader.h"
#import "Transport/TTIOTransportReader+Internal.h"
#import "Transport/TTIOTransportWriter.h"
#import "Dataset/TTIOSpectralDataset.h"
#import "Dataset/TTIOIdentification.h"
#import "Dataset/TTIOQuantification.h"
#import "Dataset/TTIOProvenanceRecord.h"
#import "Genomics/TTIOReferenceImport.h"
#import "Image/TTIOMSImage.h"
#import "HDF5/TTIOHDF5File.h"
#import "HDF5/TTIOHDF5Group.h"
#import "Providers/TTIOHDF5Provider.h"
#import "Providers/TTIOStorageProtocols.h"
#include <unistd.h>

// -------- helpers ----------------------------------------------------------

static NSString *makeTempPathP(NSString *suffix)
{
    NSString *base = [NSString stringWithFormat:@"ttio_tr_pre_%d_%@.tio",
                       (int)getpid(), suffix];
    return [NSTemporaryDirectory()
        stringByAppendingPathComponent:base];
}

// Build a .tio with exactly one reference embedded and nothing else.
// The reference has both a small chromosome (below the 4 KiB zlib
// threshold) and a large chromosome (above it) so both encoding
// branches in -writeReferenceGroup: get exercised.
static NSString *buildReferenceOnlyTio(NSString *suffix, NSError **error)
{
    NSString *path = makeTempPathP(suffix);
    unlink([path fileSystemRepresentation]);

    BOOL ok = [TTIOSpectralDataset writeMinimalToPath:path
                                                  title:@"ref-only"
                                     isaInvestigationId:@""
                                                 msRuns:@{}
                                        identifications:nil
                                        quantifications:nil
                                      provenanceRecords:nil
                                                  error:error];
    if (!ok) return nil;

    // Construct the reference and embed via the writeToDataset trick
    // (object_setIvar to attach a writable provider). Pattern from
    // TTIOReferenceImportWriteToDatasetTests.m.
    NSData *small = [@"ACGTACGT" dataUsingEncoding:NSASCIIStringEncoding];
    NSMutableData *big = [NSMutableData dataWithLength:8192];
    uint8_t *bp = (uint8_t *)big.mutableBytes;
    const char alphabet[4] = {'A', 'C', 'G', 'T'};
    for (NSUInteger i = 0; i < big.length; i++) bp[i] = alphabet[i & 3];
    TTIOReferenceImport *ref =
        [[TTIOReferenceImport alloc] initWithUri:@"prelude-ref-v1"
                                     chromosomes:@[@"chrA", @"chrB"]
                                       sequences:@[small, big]];

    TTIOHDF5Provider *p = [[TTIOHDF5Provider alloc] init];
    if (![p openURL:path mode:TTIOStorageOpenModeReadWrite error:error]) {
        return nil;
    }
    TTIOSpectralDataset *ds =
        [[TTIOSpectralDataset alloc] initWithTitle:@"ref-only"
                                isaInvestigationId:@""
                                            msRuns:@{}
                                           nmrRuns:@{}
                                   identifications:@[]
                                   quantifications:@[]
                                 provenanceRecords:@[]
                                       transitions:nil];
    Ivar provIvar = class_getInstanceVariable([TTIOSpectralDataset class],
                                                "_provider");
    if (provIvar == NULL) return nil;
    object_setIvar(ds, provIvar, p);
    if (![ref writeToDataset:ds error:error]) {
        [ds closeFile];
        return nil;
    }
    [ds closeFile];
    return path;
}

// -------- 1. reference-only round-trip via writeDataset --------------------

// Regression for the silent-drop bug: writeDataset on a reference-only
// .tio must produce a .tis that round-trips losslessly back to .tio
// with the references intact. Before Task 3.9 the references were
// dropped because writeDataset emitted nothing for the references
// accessor.
static void testReferenceOnlyRoundTripsThroughWriteDataset(void)
{
    NSError *err = nil;
    NSString *src = buildReferenceOnlyTio(@"refonly_src", &err);
    PASS(src != nil && err == nil,
         "3.9 ref-rt: built reference-only fixture .tio");

    TTIOSpectralDataset *ds =
        [TTIOSpectralDataset readFromFilePath:src error:&err];
    PASS(ds != nil && ds.references.count == 1,
         "3.9 ref-rt: fixture dataset surfaces 1 reference on open");

    NSMutableData *buf = [NSMutableData data];
    TTIOTransportWriter *w =
        [[TTIOTransportWriter alloc] initWithMutableData:buf];
    PASS([w writeDataset:ds error:&err],
         "3.9 ref-rt: writeDataset on reference-only .tio");
    [ds closeFile];

    // Floor check: a silent-drop run would have produced just a
    // StreamHeader + EndOfStream (a couple hundred bytes). 8 KiB ACGT
    // compresses to ~40 bytes via zlib, so we gate on a conservative
    // 300-byte floor.
    PASS(buf.length > 300,
         "3.9 ref-rt: stream is non-trivial (>300 bytes, "
         "rules out silent drop)");

    NSString *outPath = makeTempPathP(@"refonly_rt");
    unlink([outPath fileSystemRepresentation]);
    TTIOTransportReader *r =
        [[TTIOTransportReader alloc] initWithData:buf];
    err = nil;
    PASS([r writeTtioToPath:outPath error:&err] && err == nil,
         "3.9 ref-rt: writeTtioToPath materialised");

    TTIOSpectralDataset *back =
        [TTIOSpectralDataset readFromFilePath:outPath error:&err];
    PASS(back != nil, "3.9 ref-rt: re-opened materialised .tio");
    PASS(back.references.count == 1,
         "3.9 ref-rt: references round-tripped (count=1)");
    TTIOReferenceImport *origRef = ds.references[@"prelude-ref-v1"];
    TTIOReferenceImport *backRef = back.references[@"prelude-ref-v1"];
    PASS(backRef != nil,
         "3.9 ref-rt: back-side reference looked up by URI");
    if (origRef && backRef) {
        PASS([backRef.chromosomes isEqualToArray:origRef.chromosomes],
             "3.9 ref-rt: chromosome names round-trip");
        PASS(backRef.totalBases == origRef.totalBases,
             "3.9 ref-rt: total bases round-trip");
        PASS([backRef.md5 isEqualToData:origRef.md5],
             "3.9 ref-rt: MD5 round-trips byte-equal");
    }

    [back closeFile];
    unlink([outPath fileSystemRepresentation]);
    unlink([src fileSystemRepresentation]);
}

// -------- 2. feature flag set when any v0.11 content present ---------------

static void testFeatureFlagSetWhenReferencesPresent(void)
{
    NSError *err = nil;
    NSString *src = buildReferenceOnlyTio(@"flag_src", &err);
    PASS(src != nil && err == nil,
         "3.9 flag: built reference-only fixture .tio");

    TTIOSpectralDataset *ds =
        [TTIOSpectralDataset readFromFilePath:src error:&err];
    PASS(ds != nil, "3.9 flag: opened fixture dataset");

    NSMutableData *buf = [NSMutableData data];
    TTIOTransportWriter *w =
        [[TTIOTransportWriter alloc] initWithMutableData:buf];
    PASS([w writeDataset:ds error:&err],
         "3.9 flag: writeDataset emitted stream");

    TTIOTransportReader *r =
        [[TTIOTransportReader alloc] initWithData:buf];
    NSArray<TTIOTransportPacketRecord *> *records =
        [r readAllPacketsWithError:&err];
    PASS(records.count > 0,
         "3.9 flag: parsed stream into packet records");
    PASS(records[0].header.packetType
            == TTIOTransportPacketStreamHeader,
         "3.9 flag: first packet is StreamHeader");

    // The features list rides as UTF-8 length-prefixed strings inside
    // the StreamHeader payload — a substring search is sufficient
    // (matches Java + Python test patterns).
    NSString *sh = [[NSString alloc] initWithData:records[0].payload
                                          encoding:NSUTF8StringEncoding];
    PASS(sh != nil && [sh containsString:@"transport_v0_11"],
         "3.9 flag: StreamHeader carries transport_v0_11 feature flag");

    [ds closeFile];
    unlink([src fileSystemRepresentation]);
}

// -------- 3. v0.10 dataset unchanged: no flag, no v0.11 packets ------------

static void testV010DatasetEmitsNoV011FlagOrPackets(void)
{
    NSError *err = nil;
    NSString *src = makeTempPathP(@"v010");
    unlink([src fileSystemRepresentation]);
    BOOL ok = [TTIOSpectralDataset writeMinimalToPath:src
                                                  title:@"v010-only"
                                     isaInvestigationId:@""
                                                 msRuns:@{}
                                        identifications:nil
                                        quantifications:nil
                                      provenanceRecords:nil
                                                  error:&err];
    PASS(ok, "3.9 v010: minimal .tio with no v0.11 content");

    TTIOSpectralDataset *ds =
        [TTIOSpectralDataset readFromFilePath:src error:&err];
    PASS(ds != nil, "3.9 v010: opened v0.10 dataset");

    NSMutableData *buf = [NSMutableData data];
    TTIOTransportWriter *w =
        [[TTIOTransportWriter alloc] initWithMutableData:buf];
    PASS([w writeDataset:ds error:&err],
         "3.9 v010: writeDataset on v0.10-only fixture");

    TTIOTransportReader *r =
        [[TTIOTransportReader alloc] initWithData:buf];
    NSArray<TTIOTransportPacketRecord *> *records =
        [r readAllPacketsWithError:&err];
    PASS(records.count > 0, "3.9 v010: parsed stream");
    PASS(records[0].header.packetType
            == TTIOTransportPacketStreamHeader,
         "3.9 v010: first packet is StreamHeader");

    NSString *sh = [[NSString alloc] initWithData:records[0].payload
                                          encoding:NSUTF8StringEncoding];
    PASS(sh != nil && ![sh containsString:@"transport_v0_11"],
         "3.9 v010: v0.10-only dataset does NOT advertise transport_v0_11");

    // None of the v0.11 packet types may appear on a v0.10 stream.
    NSSet<NSNumber *> *v011Types = [NSSet setWithArray:@[
        @(TTIOTransportPacketReferenceGroupHeader),
        @(TTIOTransportPacketReferenceChromosome),
        @(TTIOTransportPacketEndOfReferenceGroup),
        @(TTIOTransportPacketImageHeader),
        @(TTIOTransportPacketImagePixel),
        @(TTIOTransportPacketEndOfImage),
        @(TTIOTransportPacketIdentificationsTable),
        @(TTIOTransportPacketQuantificationsTable),
        @(TTIOTransportPacketDatasetProvenance),
        @(TTIOTransportPacketEncryptionAlgorithm),
    ]];
    BOOL sawAny = NO;
    for (TTIOTransportPacketRecord *rec in records) {
        if ([v011Types containsObject:@(rec.header.packetType)]) {
            sawAny = YES;
            break;
        }
    }
    PASS(!sawAny,
         "3.9 v010: v0.10 stream contains zero v0.11 packets");

    [ds closeFile];
    unlink([src fileSystemRepresentation]);
}

// -------- 4. multi-section: all 6 accessors -> strict §5.4 order -----------

// Builds a .tio carrying every v0.11 first-class accessor: encryption
// algorithm root attr, dataset_provenance, one reference, an image
// cube, identifications, and quantifications. Verifies the writer
// emits them in §5.4 order on the wire.
static NSString *buildMultiSectionTio(NSError **error)
{
    NSString *path = makeTempPathP(@"multi");
    unlink([path fileSystemRepresentation]);

    // Start with an image-bearing dataset so the on-disk layout
    // includes /study/image_cube/. TTIOMSImage -writeToFilePath:
    // creates the full /study/ skeleton + the cube + axis.
    const NSUInteger w = 2, h = 2, s = 3;
    NSMutableData *cube = [NSMutableData dataWithLength:w * h * s * sizeof(double)];
    NSMutableData *mz = [NSMutableData dataWithLength:s * sizeof(double)];
    double *mzp = (double *)mz.mutableBytes;
    mzp[0] = 100.0; mzp[1] = 110.0; mzp[2] = 120.0;
    TTIOIdentification *ident = [[TTIOIdentification alloc]
        initWithRunName:@"r1"
          spectrumIndex:0
         chemicalEntity:@"CompoundA"
        confidenceScore:0.5
          evidenceChain:@[@"e1"]];
    TTIOQuantification *quant = [[TTIOQuantification alloc]
        initWithChemicalEntity:@"CompoundA"
                     sampleRef:@"s1"
                     abundance:1.0
           normalizationMethod:@"intensity-sum"
                          unit:@"counts"];
    TTIOProvenanceRecord *prov = [[TTIOProvenanceRecord alloc]
        initWithInputRefs:@[]
                 software:@"ttio-test"
               parameters:@{@"k": @"v"}
               outputRefs:@[]
            timestampUnix:12345];

    TTIOMSImage *image = [[TTIOMSImage alloc]
                            initWithTitle:@"multi-img"
                       isaInvestigationId:@""
                          identifications:@[ident]
                          quantifications:@[quant]
                        provenanceRecords:@[prov]
                                    width:w
                                   height:h
                           spectralPoints:s
                                 tileSize:32
                               pixelSizeX:10.0
                               pixelSizeY:10.0
                              scanPattern:@"raster"
                                     cube:cube
                                   mzAxis:mz];
    if (![image writeToFilePath:path error:error]) return nil;

    // Embed a reference via the same writeToDataset trick.
    NSData *small = [@"ACGT" dataUsingEncoding:NSASCIIStringEncoding];
    TTIOReferenceImport *ref =
        [[TTIOReferenceImport alloc] initWithUri:@"multi-ref-v1"
                                     chromosomes:@[@"chr1"]
                                       sequences:@[small]];
    TTIOHDF5Provider *p = [[TTIOHDF5Provider alloc] init];
    if (![p openURL:path mode:TTIOStorageOpenModeReadWrite error:error]) {
        return nil;
    }
    TTIOSpectralDataset *seed =
        [[TTIOSpectralDataset alloc] initWithTitle:@"multi"
                                isaInvestigationId:@""
                                            msRuns:@{}
                                           nmrRuns:@{}
                                   identifications:@[]
                                   quantifications:@[]
                                 provenanceRecords:@[]
                                       transitions:nil];
    Ivar provIvar = class_getInstanceVariable([TTIOSpectralDataset class],
                                                "_provider");
    if (provIvar == NULL) return nil;
    object_setIvar(seed, provIvar, p);
    if (![ref writeToDataset:seed error:error]) {
        [seed closeFile];
        return nil;
    }
    [seed closeFile];

    // Set the @encrypted root attribute so isEncrypted+encryptedAlgorithm
    // surface non-empty on open. (Encryption algorithm string only —
    // we're not exercising the cipher pipeline, just the prelude emit.)
    TTIOHDF5File *f = [TTIOHDF5File openAtPath:path error:error];
    if (!f) return nil;
    TTIOHDF5Group *root = [f rootGroup];
    if (![root setStringAttribute:@"encrypted"
                              value:@"aes-256-gcm"
                              error:error]) {
        [f close];
        return nil;
    }
    if (![f close]) return nil;
    return path;
}

static void testMultiSectionPreludeEmitsInSpec54Order(void)
{
    NSError *err = nil;
    NSString *src = buildMultiSectionTio(&err);
    PASS(src != nil && err == nil,
         "3.9 multi: built multi-section fixture .tio");

    TTIOSpectralDataset *ds =
        [TTIOSpectralDataset readFromFilePath:src error:&err];
    PASS(ds != nil, "3.9 multi: opened fixture dataset");
    // Sanity-check fixture preconditions.
    PASS(ds.isEncrypted,
         "3.9 multi: fixture precondition: isEncrypted");
    PASS(ds.provenanceRecords.count > 0,
         "3.9 multi: fixture precondition: provenance non-empty");
    PASS(ds.references.count > 0,
         "3.9 multi: fixture precondition: references non-empty");
    PASS([ds imageForKind:TTIOImageKindMS] != nil,
         "3.9 multi: fixture precondition: msImage non-nil");
    PASS(ds.identifications.count > 0,
         "3.9 multi: fixture precondition: identifications non-empty");
    PASS(ds.quantifications.count > 0,
         "3.9 multi: fixture precondition: quantifications non-empty");

    NSMutableData *buf = [NSMutableData data];
    TTIOTransportWriter *w =
        [[TTIOTransportWriter alloc] initWithMutableData:buf];
    PASS([w writeDataset:ds error:&err],
         "3.9 multi: writeDataset on multi-section .tio");

    TTIOTransportReader *r =
        [[TTIOTransportReader alloc] initWithData:buf];
    NSArray<TTIOTransportPacketRecord *> *records =
        [r readAllPacketsWithError:&err];
    PASS(records.count > 0, "3.9 multi: parsed stream");

    NSInteger encIdx = -1, provIdx = -1, refIdx = -1;
    NSInteger imgIdx = -1, idsIdx = -1, quantIdx = -1;
    for (NSUInteger i = 0; i < records.count; i++) {
        TTIOTransportPacketType t = records[i].header.packetType;
        if (t == TTIOTransportPacketEncryptionAlgorithm    && encIdx < 0)   encIdx = (NSInteger)i;
        if (t == TTIOTransportPacketDatasetProvenance      && provIdx < 0)  provIdx = (NSInteger)i;
        if (t == TTIOTransportPacketReferenceGroupHeader   && refIdx < 0)   refIdx = (NSInteger)i;
        if (t == TTIOTransportPacketImageHeader            && imgIdx < 0)   imgIdx = (NSInteger)i;
        if (t == TTIOTransportPacketIdentificationsTable   && idsIdx < 0)   idsIdx = (NSInteger)i;
        if (t == TTIOTransportPacketQuantificationsTable   && quantIdx < 0) quantIdx = (NSInteger)i;
    }
    PASS(encIdx   > 0, "3.9 multi: ENCRYPTION_ALGORITHM emitted");
    PASS(provIdx  > 0, "3.9 multi: DATASET_PROVENANCE emitted");
    PASS(refIdx   > 0, "3.9 multi: REFERENCE_GROUP_HEADER emitted");
    PASS(imgIdx   > 0, "3.9 multi: IMAGE_HEADER emitted");
    PASS(idsIdx   > 0, "3.9 multi: IDENTIFICATIONS_TABLE emitted");
    PASS(quantIdx > 0, "3.9 multi: QUANTIFICATIONS_TABLE emitted");

    PASS(encIdx < provIdx,
         "3.9 multi: §5.4: ENCRYPTION_ALGORITHM precedes "
         "DATASET_PROVENANCE");
    PASS(provIdx < refIdx,
         "3.9 multi: §5.4: DATASET_PROVENANCE precedes REFERENCE_*");
    PASS(refIdx < imgIdx,
         "3.9 multi: §5.4: REFERENCE_* precedes IMAGE_*");
    PASS(imgIdx < idsIdx,
         "3.9 multi: §5.4: IMAGE_* precedes IDENTIFICATIONS_TABLE");
    PASS(idsIdx < quantIdx,
         "3.9 multi: §5.4: IDENTIFICATIONS_TABLE precedes "
         "QUANTIFICATIONS_TABLE");

    [ds closeFile];
    unlink([src fileSystemRepresentation]);
}

// -------- entry point ------------------------------------------------------

void testTransportWriteDatasetPrelude(void);
void testTransportWriteDatasetPrelude(void)
{
    testReferenceOnlyRoundTripsThroughWriteDataset();
    testFeatureFlagSetWhenReferencesPresent();
    testV010DatasetEmitsNoV011FlagOrPackets();
    testMultiSectionPreludeEmitsInSpec54Order();
}

File: python/tests/test_transport_write_dataset_prelude.py
"""v0.11 Task 2.9 (transport-spec §5.4): end-to-end verification of
``TransportWriter.write_dataset``'s v0.11 prelude wiring.

The prelude emits in this strict order per spec §5.4:

#. ``ENCRYPTION_ALGORITHM`` (0x1B) — when ``ds.is_encrypted``
#. ``DATASET_PROVENANCE`` (0x18) — when ``ds.provenance()`` non-empty
#. ``SUBJECT_METADATA`` / ``SAMPLE_METADATA`` (0x19/0x1A) — deferred
#. Reference groups (0x10/0x11/0x12) — one sequence per ref
#. Image cube (0x13/0x14/0x15) — when ``ds.image is not None``
#. ``IDENTIFICATIONS_TABLE`` (0x16) / ``QUANTIFICATIONS_TABLE`` (0x17)

The ``transport_v0_11`` feature flag rides on ``StreamHeader.features``
when any of the above is non-empty / non-None.

Python parity for Java's
``TransportWriterReferenceWireUpTest`` (commit ``dc0de926``) plus the
post-Task-2.7 integration coverage for §5.4 ordering and the v0.10
"no v0.11 content" path.

SPDX-License-Identifier: Apache-2.0
"""
from __future__ import annotations

import io
from pathlib import Path

from ttio.enums import ImageKind
from ttio.genomic.reference_import import ReferenceImport
from ttio.identification import Identification
from ttio.provenance import ProvenanceRecord
from ttio.quantification import Quantification
from ttio.spectral_dataset import SpectralDataset
from ttio.transport.codec import TransportReader, TransportWriter
from ttio.transport.packets import PacketType, TRANSPORT_V0_11_FEATURE


def _build_reference_only(target: Path) -> Path:
    """A ``.tio`` carrying a single reference and nothing else.

    The reference is sized to BOTH straddle the 4 KiB zlib threshold
    (one chromosome below, one above) so the writer exercises both
    encoding paths in this round-trip.
    """
    SpectralDataset.write_minimal(
        target,
        title="ref_only",
        isa_investigation_id="",
        runs={},
    )
    alphabet = b"ACGT"
    big = bytes(alphabet[i & 3] for i in range(8192))
    ref = ReferenceImport(
        uri="prelude-ref-v1",
        chromosomes=["chrA", "chrB"],
        sequences=[b"ACGTACGT", big],
    )
    with SpectralDataset.open(target, writable=True) as ds_w:
        ref.write_to_dataset(ds_w)
    return target


def test_reference_only_round_trips_through_write_dataset(
    tmp_path: Path,
) -> None:
    """Regression for the silent-drop bug: ``write_dataset`` on a
    reference-only ``.tio`` must produce a ``.tis`` that round-trips
    losslessly back to ``.tio`` with the references intact. Java
    parity: ``writeDataset_round_trips_references_end_to_end``."""
    src = _build_reference_only(tmp_path / "src.tio")
    tis = tmp_path / "src.tis"
    rt = tmp_path / "rt.tio"

    with SpectralDataset.open(src) as ds:
        with open(tis, "wb") as out, TransportWriter(out) as w:
            w.write_dataset(ds)

    # The reference-only .tis must be substantially larger than the
    # known silent-drop baseline (~190 bytes — StreamHeader +
    # EndOfStream only). The 8 KiB ACGT chromosome is highly
    # compressible (zlib down to ~40 bytes), so we gate on a
    # conservative 300-byte floor: enough to fit the StreamHeader +
    # REFERENCE_GROUP_HEADER + 2x REFERENCE_CHROMOSOME +
    # END_OF_REFERENCE_GROUP + EndOfStream packet headers and short
    # bodies, but well above the silent-drop baseline.
    src_size = src.stat().st_size
    tis_size = tis.stat().st_size
    assert tis_size > 300, (
        f"reference-only write_dataset produced only {tis_size} bytes; "
        f"source was {src_size} bytes -- silent drop?"
    )

    with TransportReader(tis) as r:
        materialised = r.read_to_dataset(output_path=rt)
        materialised.close()

    with SpectralDataset.open(src) as a, SpectralDataset.open(rt) as b:
        refs_a = a.references
        refs_b = b.references
        assert list(refs_a.keys()) == list(refs_b.keys()), (
            f"reference URIs must round-trip: {list(refs_a.keys())} vs "
            f"{list(refs_b.keys())}"
        )
        for uri in refs_a.keys():
            ra = refs_a[uri]
            rb = refs_b[uri]
            assert ra.uri == rb.uri
            assert sorted(ra.chromosomes) == sorted(rb.chromosomes)
            for name in ra.chromosomes:
                assert ra.chromosome(name) == rb.chromosome(name), (
                    f"chromosome {name!r} byte content mismatch"
                )
            assert ra.total_bases == rb.total_bases
            assert ra.md5 == rb.md5


def test_feature_flag_set_when_references_present(tmp_path: Path) -> None:
    """The ``transport_v0_11`` feature flag MUST appear in
    ``StreamHeader.features`` whenever a reference rides on the
    stream (or any other v0.11 content). Java parity:
    ``writeDataset_emits_v0_11_feature_flag_when_references_present``.
    """
    src = _build_reference_only(tmp_path / "src.tio")
    out = io.BytesIO()
    with SpectralDataset.open(src) as ds:
        with TransportWriter(out) as w:
            w.write_dataset(ds)

    # Inspect the StreamHeader payload directly — the features list
    # rides as a sequence of uint16-length-prefixed UTF-8 strings, so
    # the literal flag text must appear verbatim.
    r = TransportReader(io.BytesIO(out.getvalue()))
    records = r.records_for_test()
    assert records[0].header.packet_type == int(PacketType.STREAM_HEADER), (
        "first packet must be StreamHeader"
    )
    assert TRANSPORT_V0_11_FEATURE.encode("utf-8") in records[0].payload, (
        f"StreamHeader must contain {TRANSPORT_V0_11_FEATURE!r} feature flag"
    )


def test_v0_10_dataset_emits_no_v0_11_flag_or_packets(tmp_path: Path) -> None:
    """A dataset with no v0.11 content (no references, no ids,
    no quants, no image, not encrypted, no dataset_provenance) MUST
    produce a ``.tis`` without the v0.11 feature flag and without any
    0x10-0x1B packets. Confirms the v0.10 path is unchanged."""
    src = tmp_path / "v010.tio"
    SpectralDataset.write_minimal(
        src,
        title="v010_only",
        isa_investigation_id="",
        runs={},
    )

    out = io.BytesIO()
    with SpectralDataset.open(src) as ds:
        with TransportWriter(out) as w:
            w.write_dataset(ds)

    r = TransportReader(io.BytesIO(out.getvalue()))
    records = r.records_for_test()
    # First packet must be StreamHeader and must NOT advertise the
    # v0.11 flag.
    assert records[0].header.packet_type == int(PacketType.STREAM_HEADER)
    assert TRANSPORT_V0_11_FEATURE.encode("utf-8") not in records[0].payload, (
        "v0.10-only dataset must not advertise transport_v0_11"
    )
    # None of the v0.11 packet types may appear.
    v011_types = {
        int(PacketType.REFERENCE_GROUP_HEADER),
        int(PacketType.REFERENCE_CHROMOSOME),
        int(PacketType.END_OF_REFERENCE_GROUP),
        int(PacketType.IMAGE_HEADER),
        int(PacketType.IMAGE_PIXEL),
        int(PacketType.END_OF_IMAGE),
        int(PacketType.IDENTIFICATIONS_TABLE),
        int(PacketType.QUANTIFICATIONS_TABLE),
        int(PacketType.DATASET_PROVENANCE),
        int(PacketType.SUBJECT_METADATA),
        int(PacketType.SAMPLE_METADATA),
        int(PacketType.ENCRYPTION_ALGORITHM),
    }
    for rec in records:
        assert rec.header.packet_type not in v011_types, (
            f"unexpected v0.11 packet 0x{rec.header.packet_type:02x} on "
            f"v0.10 stream"
        )


def test_multi_section_prelude_emits_in_spec_5_4_order(tmp_path: Path) -> None:
    """A dataset carrying ENCRYPTION, PROVENANCE, REFERENCE, IMAGE,
    IDENTIFICATIONS, and QUANTIFICATIONS MUST emit those sections in
    the strict §5.4 order:

      ENCRYPTION_ALGORITHM (0x1B)
      DATASET_PROVENANCE   (0x18)
      REFERENCE_*          (0x10/0x11/0x12)
      IMAGE_*              (0x13/0x14/0x15)
      IDENTIFICATIONS_TABLE (0x16)
      QUANTIFICATIONS_TABLE (0x17)

    Subjects + samples (0x19/0x1A) are deferred and skipped here.
    """
    import numpy as np
    from ttio import MSImage
    src = tmp_path / "multi.tio"
    # Build the multi-section dataset.
    image = MSImage(
        width=2, height=2, spectral_points=3,
        intensity=np.zeros((2, 2, 3), dtype=np.float64),
        mz_axis=np.array([100.0, 110.0, 120.0], dtype=np.float64),
        pixel_size_x=10.0, pixel_size_y=10.0, scan_pattern="raster",
        title="multi-img", isa_investigation_id="",
    )
    ids = [
        Identification(
            run_name="r1", spectrum_index=0,
            chemical_entity="CompoundA", confidence_score=0.5,
            evidence_chain=["e1"],
        ),
    ]
    quants = [
        Quantification(
            chemical_entity="CompoundA", sample_ref="s1",
            abundance=1.0, normalization_method="intensity-sum",
            unit="counts",
        ),
    ]
    provenance = [
        ProvenanceRecord(
            timestamp_unix=12345,
            software="ttio-test",
            parameters={"k": "v"},
            input_refs=[],
            output_refs=[],
        ),
    ]
    SpectralDataset.write_minimal(
        src,
        title="multi",
        isa_investigation_id="",
        runs={},
        identifications=ids,
        quantifications=quants,
        provenance=provenance,
        image=image,
    )
    # Add encryption attribute + a reference group post-write.
    ref = ReferenceImport(
        uri="multi-ref-v1",
        chromosomes=["chr1"],
        sequences=[b"ACGT"],
    )
    with SpectralDataset.open(src, writable=True) as ds_w:
        ref.write_to_dataset(ds_w)
        ds_w.provider.root_group().set_attribute("encrypted", "aes-256-gcm")

    out = io.BytesIO()
    with SpectralDataset.open(src) as ds:
        assert ds.is_encrypted, "fixture precondition: encryption attr set"
        assert ds.image_for_kind(ImageKind.MS) is not None
        assert ds.references
        assert ds.identifications()
        assert ds.quantifications()
        assert ds.provenance()
        with TransportWriter(out) as w:
            w.write_dataset(ds)

    r = TransportReader(io.BytesIO(out.getvalue()))
    records = r.records_for_test()

    # Index of the first packet of each v0.11 section.
    def first_idx(*types: int) -> int:
        for i, rec in enumerate(records):
            if rec.header.packet_type in types:
                return i
        return -1

    encryption_idx = first_idx(int(PacketType.ENCRYPTION_ALGORITHM))
    provenance_idx = first_idx(int(PacketType.DATASET_PROVENANCE))
    reference_idx = first_idx(int(PacketType.REFERENCE_GROUP_HEADER))
    image_idx = first_idx(int(PacketType.IMAGE_HEADER))
    ids_idx = first_idx(int(PacketType.IDENTIFICATIONS_TABLE))
    quants_idx = first_idx(int(PacketType.QUANTIFICATIONS_TABLE))

    # Every section must be present.
    assert encryption_idx > 0, "ENCRYPTION_ALGORITHM not emitted"
    assert provenance_idx > 0, "DATASET_PROVENANCE not emitted"
    assert reference_idx > 0, "REFERENCE_GROUP_HEADER not emitted"
    assert image_idx > 0, "IMAGE_HEADER not emitted"
    assert ids_idx > 0, "IDENTIFICATIONS_TABLE not emitted"
    assert quants_idx > 0, "QUANTIFICATIONS_TABLE not emitted"

    # Strict §5.4 ordering.
    assert encryption_idx < provenance_idx, (
        "§5.4 violation: ENCRYPTION_ALGORITHM must precede "
        "DATASET_PROVENANCE"
    )
    assert provenance_idx < reference_idx, (
        "§5.4 violation: DATASET_PROVENANCE must precede REFERENCE_*"
    )
    assert reference_idx < image_idx, (
        "§5.4 violation: REFERENCE_* must precede IMAGE_*"
    )
    assert image_idx < ids_idx, (
        "§5.4 violation: IMAGE_* must precede IDENTIFICATIONS_TABLE"
    )
    assert ids_idx < quants_idx, (
        "§5.4 violation: IDENTIFICATIONS_TABLE must precede "
        "QUANTIFICATIONS_TABLE"
    )
